### ISSUE TITLE
UI landing page load time regression: fetch all BenchmarkResults for the last 1000 runs when rendering landing page?

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#977]
An investigation happened as a follow-up from this conversation: https://github.com/conbench/conbench/pull/912#discussion_r1142402132

I got curious about the potential outfall of #912 in view of real-world data like the Arrow Conbench deployment. Working with hat locally (also see #976) I found a rather unambiguous performance regression for the landing page; see comment below.

**Comment by jgehrcke** [comment]
Initial impression with real-world database: rather huge impact. Got landing page load times of `0.24 s, 0.08 s` when using `lazy="select"` (not counting the results-per-run), and got e.g. `82.1 s` with `lazy="select"/"immediate"` (counting results-per-run).

**Comment by austin3dickey** [comment]
It may be worth surgically reverting this change temporarily so we can deploy without worry.

**Comment by jgehrcke** [comment]
> It may be worth surgically reverting this change temporarily so we can deploy without worry.

Yes. I want to have this resolved by Monday EOD, either with a revert or with a better solution.

I disappeared in a rabbit hole and asked some :magic_wand: magician :tophat: rabbits :rabbit: what they think about not only how to _efficiently_ query a count in a one-to-many relationship, but also how to do that in SQLAlchemy, using the new-ish primitives.

This snippet translates into a concise query and returns rather fast (below 1 second), returning both: the runs as well as the per-run result counts (but not the results, as intended):
```python
        runs_with_rescount = Session.execute(
            select(Run, func.count(BenchmarkResult.id))
            .outerjoin(BenchmarkResult)
            .group_by(Run)
            .limit(2000)
        ).all()
```

But then I hit another wall, trying to add `ORDER BY run.timestamp DESC`. I mean, I managed to add it to the query, but it resulted in the query taking infinitely long to execute. Maybe this is not even needed. To be continued! I think this might be a well-defined riddle, waiting to be solved :).

Notably, this part of SQLAlchemy talks about incorrectly-placed ORDER BY in context of a query that uses JOIN: https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#the-zen-of-joined-eager-loading

> got e.g. 82.1 s with lazy="select"/"immediate" (counting results-per-run).

Btw, the challenge here is that for the Arrow DB the newest 1000 runs are associated with almost `2*10^5` benchmark results. That's a lot of on-the-fly Python object creation (including JSON-deserialization) when queried live from the DB. For rendering the landing page, we either should not fetch those at all, or pre-fetch and serve from a cache.

**Comment by jgehrcke** [comment]
> I mean, I managed to add it to the query, but it resulted in the query taking infinitely long to execute. Maybe this is not even needed. To be continued! I think this might be a well-defined riddle, waiting to be solved :).

Those resources btw suggest that indeed it's a bit of a riddle:
- https://stackoverflow.com/questions/19057433/order-by-in-sqlalchemy-with-outer-join
- https://groups.google.com/g/sqlalchemy/c/cQ_Y2gJWj28

But I have not yet thought this through systematically. When one understands a few more things about SQL queries in general I believe this might be an easy task.

**Comment by jgehrcke** [comment]
I have for now pragmatically patches this via #1014. 

In a code comment in this patch I describe the cause for the rather big outfall. We started issuing a single SELECT statement per BenchmarkResult. In case of the Arrow Conbench database state that I was testing with, these were 164298 (yes) individual SELECT statements (i.e. 164298 benchmark results associated with the last 1000 runs).

**Comment by jgehrcke** [comment]
Before I close tabs, more related resources:
- https://sqlalchemy.narkive.com/CqRsqCLE/count-on-one-to-many-relationship
- https://datawookie.dev/blog/2021/01/sqlalchemy-efficient-counting/
- https://gist.github.com/hest/8798884

**Comment by jgehrcke** [comment]
This was resolved by #1014.
